zyte_api/utils.py

```diff
@@ -18,6 +38,100 @@ def _guess_intype(file_name, lines):
     return "txt"
 
 
+def _process_query(query):
+    """Given a query to be sent to Zyte API, return a functionally-equivalent
+    query that fixes any known issue.
+
+    Specifically, unsafe characters in the query URL are escaped.
+
+    *query* is never modified in place, but the returned object is not
+    guaranteed to be a copy of *query*: it could be *query* itself if no
+    changes where needed, or a shallow copy of *query* with some common nested
+    objects (e.g. shared ``actions`` list).
+    """
+    url = query.get("url", None)
+    if url is None:
+        return query
+    if not isinstance(url, str):
+        raise ValueError(f"Expected a str URL parameter, got {type(url)}")
+    safe_url = _safe_url_string(url)
+    if url == safe_url:
+        return query
+    return {**query, "url": safe_url}
+
+
+def _safe_url_string(
+    url: AnyStr,
+    encoding: str = "utf8",
+    path_encoding: str = "utf8",
+    quote_path: bool = True,
+) -> str:
+    """Fork of ``w3lib.url.safe_url_string`` that enforces `RFC-3986`_.
```

Comment:
hey @Gallaecio! Do you know if there is any way to prevent double-encoding in this function?

```py
In [1]: from w3lib.url import safe_download_url, safe_url_string, canonicalize_url

In [2]: safe_url_string("https://www.example.com/product/hometales-pack-of-2-sky/626857878294#bcrumbSearch:pillow%20cover")
Out[2]: 'https://www.example.com/product/hometales-pack-of-2-sky/626857878294#bcrumbSearch:pillow%20cover'

```

vs 

```py
In [1]: from zyte_api.utils import _safe_url_string

In [2]: _safe_url_string("https://www.example.com/product/hometales-pack-of-2-sky/626857878294#bcrumbSearch:pillow%20cover")
Out[2]: 'https://www.example.com/product/hometales-pack-of-2-sky/626857878294#bcrumbSearch:pillow%2520cover'
```

The issue is that sometimes input URLs are encoded, and sometimes they are not.  w3lib's `safe_url_string` tries support both. It seems that after this change we might break the client library for users who send the percent-escaped URLs (e.g. the website itself may already provide encoded versions).

Reply:
Great catch, let me fix that and add a test for it.